tests/unit/project_loader/test_preflight_check.py

```diff
@@ -0,0 +1,105 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import os
+
+import fixtures
+from testtools.matchers import Equals
+
+import snapcraft
+from snapcraft.internal.project_loader._preflight_check import conduct_preflight_check
+
+from tests import unit
+
+
+class PreflightChecksTest(unit.TestCase):
+    def setUp(self):
+        super().setUp()
+        self.fake_logger = self.useFixture(fixtures.FakeLogger(level=logging.WARN))
+
+    def assert_check_passes(self):
+        _run_check()
+        self.assertThat(self.fake_logger.output, Equals(""))
+
+    def test_no_snap_dir(self):
+        self.assert_check_passes()
+
+    def test_snapcraft_yaml(self):
+        os.makedirs("snap")
+        open(os.path.join("snap", "snapcraft.yaml"), "w").close()
+        self.assert_check_passes()
+
+    def test_global_state(self):
+        state_dir = os.path.join("snap", ".snapcraft")
+        os.makedirs(state_dir)
+        open(os.path.join(state_dir, "state"), "w").close()
+        self.assert_check_passes()
+
+    def test_icons(self):
+        gui_dir = os.path.join("snap", "gui")
+        os.makedirs(gui_dir)
+        open(os.path.join(gui_dir, "icon.png"), "w").close()
+        open(os.path.join(gui_dir, "another-icon.svg"), "w").close()
+        self.assert_check_passes()
+
+    def test_plugin(self):
+        plugins_dir = os.path.join("snap", "plugins")
+        os.makedirs(plugins_dir)
+        open(os.path.join(plugins_dir, "plugin1.py"), "w").close()
+        open(os.path.join(plugins_dir, "data-file"), "w").close()
+        self.assert_check_passes()
+
+    def test_unexpected_things(self):
+        dir1 = os.path.join("snap", "dir1")
+        dir2 = os.path.join("snap", "dir2")
+        gui_dir = os.path.join("snap", "gui")
+        state_dir = os.path.join("snap", ".snapcraft")
+        fake_plugins_dir = os.path.join(dir1, "plugins")
+        fake_gui_dir = os.path.join(dir2, "gui")
+        os.makedirs(dir1)
+        os.makedirs(dir2)
+        os.makedirs(gui_dir)
+        os.makedirs(state_dir)
+        os.makedirs(fake_plugins_dir)
+        os.makedirs(fake_gui_dir)
+        open(os.path.join(dir1, "foo"), "w").close()
+        open(os.path.join(dir2, "bar"), "w").close()
+        open(os.path.join(gui_dir, "icon.jpg"), "w").close()
+        open(os.path.join(state_dir, "baz"), "w").close()
+
+        _run_check()
+        self.assertThat(
+            self.fake_logger.output,
+            Equals(
+                "The snap/ directory is meant specifically for snapcraft, but it "
+                "contains the following non-snapcraft-related paths, which is "
+                "unsupported and will cause unexpected behavior:\n"
+                "- .snapcraft/baz\n"
+                "- dir1\n"
+                "- dir1/foo\n"
+                "- dir1/plugins\n"
+                "- dir2\n"
+                "- dir2/bar\n"
+                "- dir2/gui\n"
```

Comment:
this and the plugins one are the checks I would of asked for as regex's are easy to mess up, good anticipating there :-)